platform/engineering/design-docs/micro-frontend-architecture.md

```diff
@@ -0,0 +1,255 @@
+**Last Updated:** February XX, 2020  
+**Status:** **Draft** | In Review | Approved  
+**Approvers:** Andrew Gunsch, Rian Fowler, Wyatt Walter
+
+<!-- Should the approvers include anybody from DEPO? -->
+
+
+## Overview
+
+### Objective
+To establish a micro front end architecture which enables application teams to
+operate independently.
+
+This document is intended for front end and DevOps engineers on the
+Veteran-facing Services Platform (VSP) and Veteran-Facing Services (VSA) teams.
+
+
+### Background
+
+<!-- Could put a definitions section here if it becomes useful -->
+
+#### The State of `vets-website`
+
+##### Build
+The `vets-website` build is responsible for a lot of things. The most important
+pieces include:
+- Pulling content from the [CMS](#CMS)
+- Building that content into HTML pages
+- Validating the content for broken links, accessibility violations, etc.
+- Building all JavaScript bundles
+  - Both the common bundle and individual application bundles
+- Building all CSS bundles
+  - Both common and application-specific
+- Running all >5k unit tests for all applications
+- Running all end-to-end tests for all applications
+
+
+##### Deploy
+When `vets-website` is deployed, all content and applications are deployed
+together. **There is no way to deploy one application without deploying
+everything.**
+
+There is, however, a way to deploy the latest content from the CMS without
+deploying the latest application code. This is called a **content-only deploy**.
+The latest `vets-website` release is used instead of building code in the latest
+commit of the `vets-website` `master` branch.
+
+
+#### The Problems
+##### Failures in one Service affect all other Services
+When one application team introduces test failures into `master`, all PRs from
+all teams that include the offending commit in `master` will fail in Continuous
+Integration (CI) because all tests are run, not just the tests for a single
+application.
+
+**Example:** Team A introduces a test failure in `master`. Team B has an urgent
+bug fix that needs deployed as soon as possible, so they make a branch of
+`master` and quickly fix it. Team B's branch can't be merged until Team A's
+unrelated test failure is fixed.
+
+Likewise, when there's a content issue, such as an accessibility violation, all
+application teams' PRs will fail in CI until the accessibility violation has
+been resolved by a content editor.
+
+**Important exception:** Test failures in the latest `master` branch will not prevent
+content-only deployments since those use the latest stable `vets-website`
+release.
+
+
+###### Test failures in `master`? How?
+This happens most frequently when two individual branches work on similar parts
+of the code. The tests on each branch fail, but the combined changes cause the
+result—`master`—to fail.
+
+**TODO:** Use an actual example of where this happened.
+
+
+##### Developing applications requires CMS content
```

Comment:
how does MFE address this?

Reply:
Each application will have its own build process which creates its HTML, JavaScript, and CSS files. This would be independent from the content build process.

:thinking: Guess I didn't call that out explicitly. Just implied it with "its own CI pipeline." :+1:

Reply:
is this scoping the problem away from the content builds?

Reply:
I'm not sure I understand your question, @rianfowler.